agent/consul/stream/event_publisher.go

```diff
@@ -249,6 +248,17 @@ func (s *subscriptions) unsubscribe(req *SubscribeRequest) func() {
 	}
 }
 
+func (s *subscriptions) closeAll() {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for _, byRequest := range s.byToken {
+		for _, sub := range byRequest {
+			sub.forceClose()
```

Comment:
This just closes the stream right back to the client right? It doesn't reset their state.

Do we still have a TODO tracking full resets in the case the state store is actually being abandoned because of a Restore? In that case just closing the conn is not enough? I think we talked about fixing that later.

Reply:
This is the same type of close as an ACL token change. The client sees an error telling them to reset their state.

To allow for a "close and reconnect" we would need some kind of shared unique identifier for the state so that we can detect if a restore happened while disconnected. I created https://github.com/hashicorp/consul/issues/8354 to track that. I suspect we can leave that until later.

Reply:
Ah OK great. Leaving that till later is good.

I made a bad assumption here due to the doc comment on forceClose. I pretty sure I wrote it but it's certainly not obvious now that it is causing a full client reset now - should we consider updating that somewhere?

Reply:
I made a note to myself to improve the docs on subscription.forceClose and the related errors to include "clients must reset their state". Currently any close of the subscription will require the client to reset the state.